tests/python/gaia-ui-tests/gaiatest/gaia_test.py

```diff
@@ -739,26 +744,41 @@ def is_screen_enabled(self):
         return ret
 
     def touch_home_button(self):
+        from gaiatest.apps.lockscreen.app import LockScreen
+        from gaiatest.apps.homescreen.app import Homescreen
+
         apps = GaiaApps(self.marionette)
-        if apps.displayed_app.name.lower() != 'homescreen':
-            # touching home button will return to homescreen
-            self._dispatch_home_button_event()
-            Wait(self.marionette).until(
-                lambda m: apps.displayed_app.name.lower() == 'homescreen')
-            apps.switch_to_displayed_app()
+        apps.switch_to_displayed_app()
+
+        if apps.displayed_app.origin == Homescreen.origin:
+            self._handle_home_button_event_in_homescreen()
+        elif apps.displayed_app.origin == LockScreen.origin:
+            self._handle_home_button_event_in_lockscreen()
         else:
-            apps.switch_to_displayed_app()
-            mode = self.marionette.find_element(By.TAG_NAME, 'body').get_attribute('class')
-            self._dispatch_home_button_event()
-            apps.switch_to_displayed_app()
-            if 'edit-mode' in mode:
-                # touching home button will exit edit mode
-                Wait(self.marionette).until(lambda m: m.find_element(
-                    By.TAG_NAME, 'body').get_attribute('class') != mode)
-            else:
-                # touching home button inside homescreen will scroll it to the top
-                Wait(self.marionette).until(lambda m: m.execute_script(
-                    "return window.wrappedJSObject.scrollY") == 0)
+            self._return_to_homescreen(apps)
+
+    def _return_to_homescreen(self, apps):
+        from gaiatest.apps.homescreen.app import Homescreen
+
+        self._dispatch_home_button_event()
+        Wait(self.marionette).until(lambda m: apps.displayed_app.origin == Homescreen.origin)
+        apps.switch_to_displayed_app()
+
+    def _handle_home_button_event_in_homescreen(self):
```

Comment:
It seems to me this method should be called _dispatch_home_button_event_in_homescreen, because that's what this method is doing, dispatching the home button event. It's not responding to the home button event, it's actually dispatching it.